### DISCUSSION TITLE
Using Paths alongside a Solid object to move a platform around in a specific way?

### DISCUSSION THREAD

**Discussion by ManeaterMildred** [TrianglyRU/OrbinautFramework#48] [Q&A] [answered]
Hello! Title should be pretty explicit- I'm wondering about using Paths along with a Solid to make them move in a very specific way.

https://github.com/user-attachments/assets/f5118f30-714b-40a9-a8dc-f74891d40b9e

Unfortunately, the player does not follow along the platform in this way, and while I tried looking at the already-existing platforms for example, trying to use a Path on those leads to a similar problem, except now the graphics are also escaping-

https://github.com/user-attachments/assets/2e3d037a-ef0d-47d7-a85f-9b7445f43961

This is the added line of code for starting the path I made in obj_platform_basic.
![image_2024-11-05_040555918](https://github.com/user-attachments/assets/6f3db654-831f-4574-8367-ebef11191255)

So yeah, just wonderin' if there's a way to make this work! Sorry if the answer is obvious, I've been a bit stumped on this.

**Comment by TrianglyRU** [comment]
I've never used Paths but I think it has something to do with the code execution order and how the Path system works. You need to make sure objects moves along the path first, then obj_act_solid() is executed. I'd be grateful if you attach your project file so I can take a look whenever I have time to deal with this kind of stuff

**Reply by ManeaterMildred** [reply]
Sure thing, here: https://drive.google.com/file/d/1UGTnJxAbh5vYwAr778iV5g03Jc-b6W2L/view?usp=sharing
(I have it on restricted so other peeps don't randomly try to download it, you'll have to ask for access- Would've send a DM but alas, Github doesn't allow for DMs lmao.)

The obj_collision_move I duplicated to use the path is in Objects>Common, as usual.

**Reply by TrianglyRU** [reply]
I sent the request

**Reply by ManeaterMildred** [reply]
Accepted! Thanks for giving it a look, haha.

**Reply by TrianglyRU** [reply, marked as answer]
Okay, so, according to the GameMaker docs:
> When following a path, an instance's position is updated [after the Step event](https://manual.gamemaker.io/monthly/en/The_Asset_Editors/Object_Properties/Event_Order.htm), so during the Step event its [x](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Asset_Management/Instances/Instance_Variables/x.htm) and [y](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Asset_Management/Instances/Instance_Variables/y.htm) position will be the same as its [xprevious](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Asset_Management/Instances/Instance_Variables/xprevious.htm) and [yprevious](https://manual.gamemaker.io/monthly/en/GameMaker_Language/GML_Reference/Asset_Management/Instances/Instance_Variables/yprevious.htm) (unless manually modified in the Begin Step or Step events).

This means, we should do a workaround. Simply using End Step event won't work correctly due to how GameMaker executes events between objects, TLDR; we need to make sure it is somehow run after the Step Event but before `c_framework`'s End Step Event

To do this, open `c_framework`, End Step, and add the code below right before the `with obj_player` loop (so it will be basically the very first thing End Step will do there):
```
with c_object
{
	event_perform(ev_other, ev_user10);
}
```

Now for your object, use User Event 10 instead of Step Event. It should work just fine

**Reply by ManeaterMildred** [reply]
Hm, doesn't seem to work.

![image](https://github.com/user-attachments/assets/33816bdb-5ec3-4740-a5ba-a9b1853fee30)
![image](https://github.com/user-attachments/assets/f8fc4fad-44b8-4a3f-abc4-b9cc94b10e88)
![image](https://github.com/user-attachments/assets/a0fa2eb4-1ca0-412c-b9ec-ef6b6aaa9423)

Even with path_start called in User Event 10 and with the condition fulfilled, the block won't move.
![image](https://github.com/user-attachments/assets/03b8e79e-c3d6-41ad-9d8e-8c3524d5f81d)

Honestly, if paths just turn out to be incompatible with the framework, that's totally fine, I'll just have to find a different method of doing path movement like that, but it is odd It's not working.

**Reply by TrianglyRU** [reply]
I'd not reply with my solution if I didn't test it first. Your implementation is simply incorrect.

`c_framework`'s End Step should look like this:
![изображение](https://github.com/user-attachments/assets/0056fb63-8f60-406c-b6f8-ae4500cb8015)

For your object, you should move EVERYTHING from Step Event into User Event 10. Step Event should not be used in this case

**Reply by ManeaterMildred** [reply]
Oh I misread, apologies, that's totally on me. Yeah, I just tested it out after fixing the end_step and now it works just fine!

Thank you so much and sorry for that mishap, I somehow didn't realize I messed put the end_step setup and thought I just had to add the event_perform 10 on obj_player too, which... When I think about it for two seconds, doesn't make sense- Hah.

Thanks, again! Works just fine now-

**Reply by TrianglyRU** [reply]
Glad to help!
